File: 2024/4/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
)

func readData() ([]string, error) {
	file, err := os.Open(filepath.Join(".", "data.txt"))
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var data []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		data = append(data, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return data, nil
}

func formatData(rows []string) [][]rune {
	var grid [][]rune
	for _, row := range rows {
		grid = append(grid, []rune(row))
	}
	return grid
}

func part1(grid [][]rune) int {
	count := 0
	directions := [][]int{
		{-1, -1}, {-1, 0}, {-1, 1},
		{0, 1}, {0, -1},
		{1, -1}, {1, 0}, {1, 1},
	}

	for row := 0; row < len(grid); row++ {
		for col := 0; col < len(grid[0]); col++ {
			for _, direction := range directions {
				if dfs(grid, row, col, direction, "XMAS") {
					count++
				}
			}
		}
	}

	fmt.Println(count)
	return count
}

func dfs(grid [][]rune, row, col int, direction []int, word string) bool {
	r, c := row, col
	for _, letter := range word {
		if r < 0 || r >= len(grid) || c < 0 || c >= len(grid[0]) {
			return false
		}
		if grid[r][c] != letter {
			return false
		}
		r += direction[0]
		c += direction[1]
	}
	return true
}

func part2(grid [][]rune) int {
	count := 0
	directions := [][]int{
		{1, 1},  // right down
		{1, -1}, // left down
	}

	for row := 0; row < len(grid); row++ {
		for col := 0; col < len(grid[0]); col++ {
			if grid[row][col] != 'M' && grid[row][col] != 'S' {
				continue
			}

			word := "MAS"
			if grid[row][col] == 'S' {
				word = "SAM"
			}
			middle := len(word) / 2

			if col+len(word)-1 >= len(grid[0]) {
				continue
			}

			crossingWord := "MAS"
			if grid[row][col+len(word)-1] == 'S' {
				crossingWord = "SAM"
			}

			if dfs(grid, row, col, directions[0], word) &&
				dfs(grid, row, col+len(word)-1, directions[1], crossingWord) &&
				grid[row+middle][col+middle] == rune(word[middle]) {
				count++
			}
		}
	}

	fmt.Println(count)
	return count
}

func main() {
	data, err := readData()
	if err != nil {
		fmt.Println("Get rekt:", err)
		return
	}

	formattedData := formatData(data)
	part1(formattedData)
	part2(formattedData)
}

File: 2024/4/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn read_data() -> io::Result<Vec<String>> {
    let path = Path::new("data.txt");
    let file = File::open(&path)?;
    let reader = io::BufReader::new(file);

    let mut data = Vec::new();
    for line in reader.lines() {
        data.push(line?);
    }

    Ok(data)
}

fn format_data(rows: Vec<String>) -> Vec<Vec<char>> {
    rows.into_iter().map(|row| row.chars().collect()).collect()
}

fn part1(grid: &[Vec<char>]) -> usize {
    let mut count = 0;
    let directions = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, 1), (0, -1),
        (1, -1), (1, 0), (1, 1),
    ];

    for row in 0..grid.len() {
        for col in 0..grid[0].len() {
            for direction in &directions {
                if dfs(grid, row as isize, col as isize, *direction, "XMAS") {
                    count += 1;
                }
            }
        }
    }

    println!("{}", count);
    count
}

fn dfs(grid: &[Vec<char>], mut row: isize, mut col: isize, direction: (isize, isize), word: &str) -> bool {
    for letter in word.chars() {
        if row < 0 || row >= grid.len() as isize || col < 0 || col >= grid[0].len() as isize {
            return false;
        }
        if grid[row as usize][col as usize] != letter {
            return false;
        }
        row += direction.0;
        col += direction.1;
    }
    true
}

fn part2(grid: &[Vec<char>]) -> usize {
    let mut count = 0;
    let directions = [
        (1, 1),  // right down
        (1, -1), // left down
    ];

    for row in 0..grid.len() {
        for col in 0..grid[0].len() {
            if grid[row][col] != 'M' && grid[row][col] != 'S' {
                continue;
            }

            let word = if grid[row][col] == 'M' { "MAS" } else { "SAM" };
            let middle = word.len() / 2;

            if col + word.len() - 1 >= grid[0].len() {
                continue;
            }

            let crossing_word = if grid[row][col + word.len() - 1] == 'M' { "MAS" } else { "SAM" };

            if dfs(grid, row as isize, col as isize, directions[0], word) &&
                dfs(grid, row as isize, (col + word.len() - 1) as isize, directions[1], crossing_word) &&
                grid[row + middle][col + middle] == word.chars().nth(middle).unwrap() {
                count += 1;
            }
        }
    }

    println!("{}", count);
    count
}

fn main() -> io::Result<()> {
    let data = read_data()?;
    let formatted_data = format_data(data);
    part1(&formatted_data);
    part2(&formatted_data);
    Ok(())
}